accepted/2023/wasm-browser-threads.md

```diff
@@ -0,0 +1,506 @@
+- no user code JS interop on any thread
+- internal solutions for Blazor needs
+- Ignores problems **1,2,3,4,5**
+
+## (10) Sidecar + only async just JS proxies UI + JSWebWorker + Blazor WASM server
+- **C,E,H,N,P,S,U,W+Y,c,e+f,h+k,m,o,q,w**
+- no C or managed code on UI thread
+- no support for blocking sync JSExport calls from UI thread (callbacks)
+    - it will throw PNSE
+- this will create double proxy for `Task`, `JSObject`, `Func<>` etc
+    - difficult to GC, difficult to debug
+- double marshaling of parameters
+- Solves **1,2** for managed code.
+- Avoids **1,2** for JS callback
+    - emscripten main loop stays responsive
+- Solves **3,4,5**
+
+## (11) Sidecar + async & sync just JS proxies UI + JSWebWorker + Blazor WASM server
+- **C,F,H,N,P,S,U,W+Y,c,e+f,h+k,m,o,q,w**
+- no C or managed code on UI thread
+- support for blocking sync JSExport calls from UI thread (callbacks)
+    - at blocking the UI is at least well isolated from runtime code
+    - it makes responsibility for sync call clear
+- this will create double proxy for `Task`, `JSObject`, `Func<>` etc
+    - difficult to GC, difficult to debug
+- double marshaling of parameters
+- Solves **1,2** for managed code
+    - unless there is sync `JSImport`->`JSExport` call
+- Ignores **1,2** for JS callback
+    - emscripten main loop stays responsive
+- Solves **3,4,5**
+
+## (12) Deputy + managed dispatch to UI + JSWebWorker + with sync JSExport
+- **B,F,K,N,Q,S/T,U,W,a/b/c,d+f,h,l,n,s/z,v**
+- this uses `JSSynchronizationContext` to dispatch calls to UI thread
+    - this is "dirty" as compared to sidecar because some managed code is actually running on UI thread
+    - it needs to also use `SynchronizationContext` for `JSExport` and callbacks, to dispatch to deputy.
+- blazor render could be both legacy render or Blazor server style
+    - because we have both memory and mono on the UI thread
+- Solves **1,2** for managed code
+    - unless there is sync `JSImport`->`JSExport` call
+- Ignores **1,2** for JS callback
+    - emscripten main loop could deadlock on sync JSExport
+- Solves **3,4,5**
+
+## (13) Deputy + emscripten dispatch to UI + JSWebWorker + with sync JSExport
+- **B,F,J,N,R,T,U,W,a/b/c,d+f,i,l,n,s,v**
+- is variation of **(12)**
+    - with emscripten dispatch and marshaling in UI thread
+- this uses `emscripten_dispatch_to_thread_async` for `call_entry_point`, `complete_task`, `cwraps.mono_wasm_invoke_method_bound`, `mono_wasm_invoke_bound_function`, `mono_wasm_invoke_import`, `call_delegate_method` to get to the UI thread.
+- it uses other `cwraps` locally on UI thread, like `mono_wasm_new_root`, `stringToMonoStringRoot`, `malloc`, `free`, `create_task_callback_method`
+    - it means that interop related managed runtime code is running on the UI thread, but not the user code.
+    - it means that parameter marshalling is fast (compared to sidecar)
+    - it still needs to enter GC barrier and so it could block UI for GC run shortly
```

Comment:
An interesting idea here is to create a GC-aware "transactions" API between the interop layer and the runtime.  Essentially: allow the marshaling code to succeed at calls to the runtime even if the GC is currently in a STW pause.  But instead of actually doing a GC Unsafe transition and executing some operation, just create a transaction log entry and immediately return back to marshaling.  Then after all the operations are queued up, the marshaling could commit the transaction and this will either block until GC STW is finished, or immediately do the operation.
(Another variant of this: if the marshaling is just interacting with handles or proxy objects and doesn't actually care about the final locations of things in the managed memory - just allocate into some side buffer and queue some work to be done at the end of STW to promote the values from the side buffer to the real managed heap).

This could possibly be done as a later optimization...

Reply:
This is good idea.

I think that proxy disposal could be scheduled to later time.

I hope we are able to schedule also proxy **creation** via pre-existing handle, it would make sidecar approach more performant. 
Because it would not have to do per parameter call cross-threads. The difficult one I can think of would be 
`GCHandle` of a `TaskCompletionSource`. We could pre-allocate some TCSs and then just use them as needed.

If we are also able to schedule allocation (for arrays), perhaps we could serialize it all to the point that we could avoid double proxy of "comlink".

Reply:
Well, not so easy we can't pre-create arbitrary `TaskCompletionSource` because we are strongly typed `TaskCompletionSource<T>` in many cases.

Reply:
I added this as design **(16)**